cocos/scripting/js-bindings/manual/jsb_global.cpp

```diff
@@ -4,27 +4,83 @@
 
 #include "jsb_global.h"
 #include "jsb_conversions.hpp"
+#include "xxtea/xxtea.h"
 
 using namespace cocos2d;
 
 se::Object* __jscObj = nullptr;
 se::Object* __ccObj = nullptr;
 se::Object* __jsbObj = nullptr;
 
+const char* BYTE_CODE_FILE_EXT = ".jsc";
+
+static std::string xxteaKey = "";
+
+void jsb_set_xxtea_key(const std::string& key)
+{
+    xxteaKey = key;
+}
+
+static std::string removeFileExt(const std::string& filePath)
+{
+    size_t pos = filePath.rfind('.');
+    if (0 < pos)
+    {
+        return filePath.substr(0, pos);
+    }
+    return filePath;
+}
+
 bool jsb_run_script(const std::string& filePath)
 {
     static se::ScriptEngine::FileOperationDelegate delegate;
     if (!delegate.isValid())
     {
         delegate.onGetDataFromFile = [](const std::string& path, const uint8_t** data, size_t* dataLen) -> void{
             assert(!path.empty() && data != nullptr && dataLen != nullptr);
-            Data fileData = FileUtils::getInstance()->getDataFromFile(path);
+            
+            Data fileData;
+            
+#if SCRIPT_ENGINE_TYPE != SCRIPT_ENGINE_SM
+            std::string byteCodePath = removeFileExt(path) + BYTE_CODE_FILE_EXT;
+            if (FileUtils::getInstance()->isFileExist(byteCodePath)) {
+                fileData = FileUtils::getInstance()->getDataFromFile(byteCodePath);
+                
+                *data = xxtea_decrypt((unsigned char*)fileData.getBytes(), (uint32_t)fileData.getSize(), (unsigned char*)xxteaKey.c_str(), (uint32_t)xxteaKey.size(), (uint32_t*)&dataLen);
+                
+                if (data == nullptr) {
+                    SE_REPORT_ERROR("Can't decrypt code for %s", byteCodePath.c_str());
+                }
+                
+                return;
+            }
+            
+#endif
+            fileData = FileUtils::getInstance()->getDataFromFile(path);
             *data = fileData.getBytes();
             *dataLen = fileData.getSize();
         };
 
         delegate.onGetStringFromFile = [](const std::string& path) -> std::string{
             assert(!path.empty());
+            
+#if SCRIPT_ENGINE_TYPE != SCRIPT_ENGINE_SM
+            std::string byteCodePath = removeFileExt(path) + BYTE_CODE_FILE_EXT;
+            if (FileUtils::getInstance()->isFileExist(byteCodePath)) {
+                Data fileData = FileUtils::getInstance()->getDataFromFile(byteCodePath);
+                
+                size_t retLength;
+                unsigned char * data = xxtea_decrypt((unsigned char*)fileData.getBytes(), (uint32_t)fileData.getSize(), (unsigned char*)xxteaKey.c_str(), (uint32_t)xxteaKey.size(), (uint32_t*)&retLength);
+                
+                if (data == nullptr) {
+                    SE_REPORT_ERROR("Can't decrypt code for %s", byteCodePath.c_str());
+                    return "";
+                }
+                
+                return std::string(reinterpret_cast<const char*>(data));
```

Comment:
Is the data with a `null` terminated letter?
It's better to use `return std::string(reinterpret_cast<const char*>(data), retLength)` to make a safe conversion to string.